### ISSUE TITLE
Custom fields (Eta support)

### ISSUE THREAD

**Issue by carymrobbins** [sol/hpack#222]
Eta projects which contain Java sources require a `java-sources:` field in the cabal file -

```cabal
library
  java-sources:
      java/eta/parser/Utils.java
```

Does hpack provide any support for custom cabal fields? That would be ideal to avoid having to PR this as a new feature. Otherwise, we would want to look into adding the `java-sources` field to component stanzas.

**Comment by tfausak** [comment]
This is only tangentially related, but: I thought custom Cabal fields were supposed to start with `x-`. Does Cabal parse packages with `java-sources` correctly?

**Comment by carymrobbins** [comment]
@tfausak Ah you are correct, when I run it through the cabal parser I get -

```
PWarning "Unknown fields: java-sources ...
```

The parse succeeds and `java-sources` is omitted from the parse tree.

**Comment by sol** [comment]
@carymrobbins What exact behavior would we want here from `hpack`?

**Comment by sol** [comment]
Any example cabal files that I could take a look at?

**Comment by carymrobbins** [comment]
@sol - Here's one for a project I'd like to transition to use hpack

https://gist.github.com/carymrobbins/77582cb511093e512bd568a26b77a011

**Comment by sol** [comment]
@carymrobbins `hpack-0.24.0` adds support a `verbatim` cabal escape hatch.  With this you can include custom fields, e.g.:

```yaml
library:
  verbatim:
    java-sources: java/eta/parser/Utils.java
```

Can you try if this works for you?

**Comment by sol** [comment]
@carymrobbins I'm closing this, but please still let me know how it goes + don't be shy to open an other issue if needed (I'll try to be more responsive in the future!).

**Comment by carymrobbins** [comment]
@sol I structure it like this I get a cryptic error -

```yaml
  verbatim:
    java-sources:
      - java/eta/parser/Utils.java
```

> package.yaml: Error while parsing $.library.verbatim - expected String or Object, encountered Object

The workaround is to structure it as you did -

```yaml
  verbatim:
    java-sources: java/eta/parser/Utils.java
```

Note that for multiple entries, you can just omit the `-` prefixes and it works well (I assume it is just interpreted as a string literal this way) -

```yaml
  verbatim:
    java-sources:
      java/eta/parser/Foo.java
      java/eta/parser/Bar.java
```

**Comment by sol** [comment]
Hey, the error message is indeed wrong.  Thanks a lot for pointing this out!  I'll see were our generic machinery chocks.

Currently, only `String`, `Number`, `Bool` and `Null` are accepted as values for the object.  Top level, to `verbatim` you can use an object or a string.

And yes, multi-line string literals are what you can use if you have a list.

**Comment by sol** [comment]
One more thing, if you need the newlines in the generated output change it to:

```yaml
  verbatim:
    java-sources: |
      java/eta/parser/Foo.java
      java/eta/parser/Bar.java
```

**Comment by sol** [comment]
Error messages are fixed in 793f131.
